ScoutSuite/providers/azure/resources/sql/databases.py

```diff
@@ -0,0 +1,69 @@
+# -*- coding: utf-8 -*-
+
+from azure.mgmt.sql import SqlManagementClient
+
+from ScoutSuite.providers.base.configs.resource_config import ResourceConfig
+
+from .database_blob_auditing_policies import DatabaseBlobAuditingPoliciesConfig
+from .database_threat_detection_policies import DatabaseThreatDetectionPoliciesConfig
+from .replication_links import ReplicationLinksConfig
+from .transparent_data_encryptions import TransparentDataEncryptionsConfig
+
+
+class DatabasesConfig(ResourceConfig):
+    # the following static methods will be used to parse the config of each resource nested in a database config
+    # (defined in 'children' attribute below):
+    @staticmethod
+    def _parse_auditing_config(config):
+        return {
+            'auditing_enabled': config.value.state == "Enabled"
+        }
+
+    @staticmethod
+    def _parse_threat_detection_config(config):
+        return {
+            'threat_detection_enabled': config.value.state == "Enabled"
+        }
+
+    @staticmethod
+    def _parse_replication_links_config(config):
+        return {
+            'replication_configured': len(config.value) > 0
+        }
+
+    @staticmethod
+    def _parse_transparent_data_encryption_config(config):
+        return {
+            'transparent_data_encryption_enabled': config.value.status == "Enabled"
+        }
+
+    # register children resources:
+    children = {
+        'Database Auditing Settings': (DatabaseBlobAuditingPoliciesConfig, _parse_auditing_config),
+        'Database Threat Detection Policies': (DatabaseThreatDetectionPoliciesConfig, _parse_threat_detection_config),
+        'Replication Links': (ReplicationLinksConfig, _parse_replication_links_config),
+        'Transparent Data Encryptions': (TransparentDataEncryptionsConfig, _parse_transparent_data_encryption_config)
+    }
+
+    def __init__(self, resource_group_name, server_name):
+        self.resource_group_name = resource_group_name
+        self.server_name = server_name
+
+        self.databases = {}
+
+    async def fetch_all(self, credentials, regions=None, partition_name='aws', targets=None):
+        # sdk container:
+        api = SqlManagementClient(credentials.credentials, credentials.subscription_id)
+
+        # TODO: for db in await api.databases.list_by_server(self.resource_group_name, self.server_name):
+        for db in api.databases.list_by_server(self.resource_group_name, self.server_name):
+            self.databases[db.name] = {
+                'id': db.name,
+            }
+
+            # put the following code in a fetch_children() parent method (typical method for a composite node)?
+            for (resource_config, resource_parser) in self.children.values():
+                resource = resource_config(self.resource_group_name, self.server_name, db.name)
+                await resource.fetch_all(credentials)
+                for k, v in resource_parser.__func__(resource).items():
+                    self.databases[db.name][k] = v
```

Comment:
True, I think this piece of code could be moved up to a `ResourceGroupComposite` class. One thing though is it seems like you're putting the children directly in the parent, which is kind of flattening the composite structure.
 
The tree looks like:
```json
"databases": {
    "some_database": {
        "auditing_enabled": true,
        "threat_detection_enabled": true,
        "replication_configured": true,
        "transparent_data_encryption_enabled": true
    },
    "... more of that"
}
```

But from my understanding, the tree should look like that

```
"databases": {
    "some_database": {
        "BlobAuditingPolicies" {
            "auditing_enabled": true
            "some_policy":  { ... },
            "some_other_policy": { ... }
         },
        "ThreatDetectionPolicies": {
            "threat_detection_enabled": true,
            "some_policy":  { ... },
            "some_other_policy": { ... }
        },
        "ReplicationLinks": {
            "replication_configured": true,
            "some_link_or_whatever...?": { "... }
            "another_link_config_iguess": { ... }
        }  
        "TransparentDataEncryptions": {
            "transparent_data_encryption_enabled": true
            "it_doesnt_even_make_sense_here": { ... }
        }
    },
    "... more of that"
}
```
Of course, this doesn't make sense. I think you went one level deeper than what I had understood. In my mind, we were going to split in different `_________Config` class every time you have an id basically. So let's say you have `ec2.regions.id.vpcs.id.instances.id`, then you have `Ec2Config`, `RegionsConfig`, `VpcsConfig`, and `InstancesConfig`. It seems like you created a config class for each attribute.

Reply:
You're correct about the trees. They are different because I'm making a difference between the structure of azure configuration (which looks like your 2nd tree, the most nested one) and the json we want to build and pass to the rule engine (which, for obvious reasons, doesn't need to store the entire azure configuration but just some relevant attributes.
That being said, I think flattening the composite structure like I do is related to your question about giving the responsability to a node to parse its children configs.

> In my mind, we were going to split in different `_________Config` class every time you have an id > basically. So let's say you have `ec2.regions.id.vpcs.id.instances.id`, then you have `Ec2Config`, > `RegionsConfig`, `VpcsConfig`, and `InstancesConfig`. It seems like you created a config class for > each attribute.

Yes, I created a config class for each attribute but I did so because those attributes have their own API call and therefore, look like "leaves" (of our composite pattern) to me. But I don't think AWS API is that much granular. Looks like only one API call `describe_instances` is necessary to retrieve all relevant attributes of an instance... So having nothing nested under `InstancesConfig` would indeed makes sense.

Reply:
I see, fair enough. The flattening still bothers me though, it kind of breaks the structure. 

Here are a few solutions

- Use some kind of `SingleConfig` abstract class to make the behavior explicit. 
- Merge the 4 deepest configs into one `DatabaseConfig` class. 

I'll continue thinking about this.

Reply:
Just updated the code to make each node take care of its own parsing. The flattening part has changed but it is still not very satisfying to me, I'll keep thinking about it.
I'll had a azure_facade tomorrow, to take care about the API calls!

Reply:
> I don't think AWS API is that much granular

Isn't it though? I think that's why there are so many callbacks used in the AWS services.

Reply:
Most of the services have "describe_instance", which allows you to get a bunch of configs. From my understanding, with Azure, it seems like you often need to make separate calls, and some calls are dependant on getting info from other calls. It has a deeper hierarchy, basically.